### PR TITLE
implement zeek dns logs parser

### PR DESCRIPTION
**Pull request by sardev08**
## Changes
Implement zeek dns logs parser

## Testing
Unit tests

### CODE REVIEW

File: internal/log_analysis/log_processor/parsers/timestamp/timestamp.go [hunk 1]
```diff
@@ -100,7 +100,16 @@ func (ts *UnixMillisecond) MarshalJSON() ([]byte, error) {
 func (ts *UnixMillisecond) UnmarshalJSON(jsonBytes []byte) (err error) {
 	value, err := strconv.ParseInt(string(jsonBytes), 10, 64)
 	if err != nil {
-		return err
+		f, err := strconv.ParseFloat(string(jsonBytes), 64)
```
Comment (rleighton) [inline internal/log_analysis/log_processor/parsers/timestamp/timestamp.go]: Please add some tests of int's and floats

File: internal/log_analysis/log_processor/parsers/timestamp/timestamp.go [hunk 2]
```diff
@@ -100,7 +100,16 @@ func (ts *UnixMillisecond) MarshalJSON() ([]byte, error) {
...
+		if err != nil {
+			return err
+		}
+
+		secs := int64(f)
+		nsecs := int64((f - float64(secs)) * 1e9)
```
Comment (rleighton) [inline internal/log_analysis/log_processor/parsers/timestamp/timestamp.go]: I think this would save a floating point multiply:
```
nsecs := int64(f - float64(secs)) * 1000000000
```

File: internal/log_analysis/log_processor/parsers/zeeklogs/dns.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	UID       *string                    `json:"uid,omitempty" validate:"required" description:"A unique identifier of the connection over which DNS messages are being transferred."`
+	IDOrigH   *string                    `json:"id.orig_h" validate:"required" description:"The originator’s IP address."`
+	IDOrigP   *int                       `json:"id.orig_p" validate:"required" description:"The originator’s port number."`
+	IDRespH   *string                    `json:"id.resp_h" validate:"required" description:"The responder’s IP address."`
+	IDRespP   *int                       `json:"id.resp_p" validate:"required" description:"The responder’s port number."`
+	Proto     *string                    `json:"proto" validate:"required" description:"The transport layer protocol of the connection."`
+	TransID   *int                       `json:"trans_id,omitempty" description:"A 16-bit identifier assigned by the program that generated the DNS query. Also used in responses to match up replies to outstanding queries."`
```
Comment (rleighton) [inline internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: *uint16 ?

File: internal/log_analysis/log_processor/parsers/zeeklogs/dns.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+Reference: https://docs.zeek.org/en/current/scripts/base/protocols/dns/main.zeek.html#type-DNS::Info`
+
+// nolint:lll
+type ZeekDNS struct {
+	Ts        *timestamp.UnixMillisecond `json:"ts,omitempty" validate:"required" description:"The earliest time at which a DNS protocol message over the associated connection is observed."`
+	UID       *string                    `json:"uid,omitempty" validate:"required" description:"A unique identifier of the connection over which DNS messages are being transferred."`
+	IDOrigH   *string                    `json:"id.orig_h" validate:"required" description:"The originator’s IP address."`
+	IDOrigP   *int                       `json:"id.orig_p" validate:"required" description:"The originator’s port number."`
+	IDRespH   *string                    `json:"id.resp_h" validate:"required" description:"The responder’s IP address."`
+	IDRespP   *int                       `json:"id.resp_p" validate:"required" description:"The responder’s port number."`
```
Comment (rleighton) [inline internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: Port should be *uint16

File: internal/log_analysis/log_processor/parsers/zeeklogs/dns.go [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+var ZeekDNSDesc = `Zeek DNS activity
+Reference: https://docs.zeek.org/en/current/scripts/base/protocols/dns/main.zeek.html#type-DNS::Info`
+
+// nolint:lll
+type ZeekDNS struct {
+	Ts        *timestamp.UnixMillisecond `json:"ts,omitempty" validate:"required" description:"The earliest time at which a DNS protocol message over the associated connection is observed."`
+	UID       *string                    `json:"uid,omitempty" validate:"required" description:"A unique identifier of the connection over which DNS messages are being transferred."`
+	IDOrigH   *string                    `json:"id.orig_h" validate:"required" description:"The originator’s IP address."`
+	IDOrigP   *int                       `json:"id.orig_p" validate:"required" description:"The originator’s port number."`
```
Comment (rleighton) [inline internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: Port should be *uint16

File: internal/log_analysis/log_processor/parsers/zeeklogs/dns.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+	AA        *bool                      `json:"AA,omitempty" description:"The Authoritative Answer bit for response messages specifies that the responding name server is an authority for the domain name in the question section."`
+	TC        *bool                      `json:"TC,omitempty" description:"The Truncation bit specifies that the message was truncated."`
+	RD        *bool                      `json:"RD,omitempty" description:"The Recursion Desired bit in a request message indicates that the client wants recursive service for this query."`
+	RA        *bool                      `json:"RA,omitempty" description:"The Recursion Available bit in a response message indicates that the name server supports recursive queries."`
+	Z         *int                       `json:"Z,omitempty" description:"A reserved field that is usually zero in queries and responses."`
+	Answers   []string                   `json:"answers,omitempty" description:"The set of resource descriptions in the query answer."`
+	TTLs      []float64                  `json:"TTLs,omitempty" description:"The caching intervals of the associated RRs described by the answers field."`
```
Comment (rleighton) [inline internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: Are these really floats?

File: internal/log_analysis/log_processor/parsers/zeeklogs/dns.go [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+	IDOrigH   *string                    `json:"id.orig_h" validate:"required" description:"The originator’s IP address."`
+	IDOrigP   *int                       `json:"id.orig_p" validate:"required" description:"The originator’s port number."`
+	IDRespH   *string                    `json:"id.resp_h" validate:"required" description:"The responder’s IP address."`
+	IDRespP   *int                       `json:"id.resp_p" validate:"required" description:"The responder’s port number."`
+	Proto     *string                    `json:"proto" validate:"required" description:"The transport layer protocol of the connection."`
+	TransID   *int                       `json:"trans_id,omitempty" description:"A 16-bit identifier assigned by the program that generated the DNS query. Also used in responses to match up replies to outstanding queries."`
+	Query     *string                    `json:"query,omitempty" description:"The domain name that is the subject of the DNS query."`
+	Rcode     *int                       `json:"rcode,omitempty" description:"The response code value in DNS response messages."`
```
Comment (rleighton) [inline internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: Is this a signed?

File: internal/log_analysis/log_processor/parsers/zeeklogs/dns.go [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+// LogType returns the log type supported by this parser
+func (p *ZeekDNSParser) LogType() string {
+	return "Zeek.DNS"
+}
+
+func (event *ZeekDNS) updatePantherFields(p *ZeekDNSParser) {
+	event.SetCoreFields(p.LogType(), (*timestamp.RFC3339)(event.Ts), event)
+	if event.IDOrigH != nil {
```
Comment (rleighton) [inline internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: You should update event.AppendAnyIPAddresses() with:
* IDRespH
* IDOrigH
* Check all Answers, and if net.ParseIP(answer) != nil, add the answer

File: internal/log_analysis/log_processor/parsers/timestamp/timestamp.go [hunk 9]
```diff
@@ -100,7 +100,16 @@ func (ts *UnixMillisecond) MarshalJSON() ([]byte, error) {
 func (ts *UnixMillisecond) UnmarshalJSON(jsonBytes []byte) (err error) {
 	value, err := strconv.ParseInt(string(jsonBytes), 10, 64)
 	if err != nil {
-		return err
+		f, err := strconv.ParseFloat(string(jsonBytes), 64)
+		if err != nil {
+			return err
+		}
+
+		secs := int64(f)
```
Comment (rleighton) [inline internal/log_analysis/log_processor/parsers/timestamp/timestamp.go]: See Jermey's comments about units

File: internal/log_analysis/log_processor/parsers/zeeklogs/dns.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+		// The hostname should be a FQDN, but may also be an IP address. Check for IP, otherwise
```
Comment (rleighton) [inline internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: hostname?

File: internal/log_analysis/log_processor/parsers/zeeklogs/dns.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+		// add as a domain name. https://tools.ietf.org/html/rfc3164#section-6.2.4
+		hostname := *event.IDOrigH
+		if net.ParseIP(hostname) != nil {
+			event.AppendAnyIPAddresses(hostname)
+		} else {
+			event.AppendAnyDomainNames(hostname)
```
Comment (rleighton) [inline internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: Can't happen. Why is this here?

File: internal/log_analysis/log_processor/parsers/zeeklogs/dns.go [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+	IDOrigH   *string              `json:"id.orig_h" validate:"required" description:"The originator’s IP address."`
+	IDOrigP   *uint16              `json:"id.orig_p" validate:"required" description:"The originator’s port number."`
+	IDRespH   *string              `json:"id.resp_h" validate:"required" description:"The responder’s IP address."`
+	IDRespP   *uint16              `json:"id.resp_p" validate:"required" description:"The responder’s port number."`
+	Proto     *string              `json:"proto" validate:"required" description:"The transport layer protocol of the connection."`
+	TransID   *uint16              `json:"trans_id,omitempty" description:"A 16-bit identifier assigned by the program that generated the DNS query. Also used in responses to match up replies to outstanding queries."`
+	Query     *string              `json:"query,omitempty" description:"The domain name that is the subject of the DNS query."`
+	Rcode     *uint64              `json:"rcode,omitempty" description:"The response code value in DNS response messages."`
```
Comment (rleighton) [inline internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: rcode is 4 bits (check this, it is from memory). Change this to *uint8

File: internal/log_analysis/log_processor/parsers/zeeklogs/dns.go [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+	AA        *bool                `json:"AA,omitempty" description:"The Authoritative Answer bit for response messages specifies that the responding name server is an authority for the domain name in the question section."`
+	TC        *bool                `json:"TC,omitempty" description:"The Truncation bit specifies that the message was truncated."`
+	RD        *bool                `json:"RD,omitempty" description:"The Recursion Desired bit in a request message indicates that the client wants recursive service for this query."`
+	RA        *bool                `json:"RA,omitempty" description:"The Recursion Available bit in a response message indicates that the name server supports recursive queries."`
+	Z         *int                 `json:"Z,omitempty" description:"A reserved field that is usually zero in queries and responses."`
+	Answers   []string             `json:"answers,omitempty" description:"The set of resource descriptions in the query answer."`
+	TTLs      []float64            `json:"TTLs,omitempty" description:"The caching intervals of the associated RRs described by the answers field."`
```
Comment (rleighton) [inline internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: Why is this a float64?

File: internal/log_analysis/log_processor/parsers/zeeklogs/dns.go (hunk 12 above)
Comment (sardev08) [reply internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: I can change it, but it uses count type and docs said that it's uint64
https://docs.zeek.org/en/current/scripts/base/protocols/dns/main.zeek.html#type-DNS::Info

File: internal/log_analysis/log_processor/parsers/zeeklogs/dns.go (hunk 13 above)
Comment (sardev08) [reply internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: according with docs and logs example it's float64
https://docs.zeek.org/en/current/scripts/base/protocols/dns/main.zeek.html#type-DNS::Info

File: internal/log_analysis/log_processor/parsers/zeeklogs/dns.go (hunk 12 above)
Comment (rleighton) [reply internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: Ok, just checking ... logs xform the raw data.

File: internal/log_analysis/log_processor/parsers/zeeklogs/dns.go (hunk 13 above)
Comment (rleighton) [reply internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: Ok, add what the units are to the description (seconds?)

File: internal/log_analysis/log_processor/parsers/zeeklogs/dns.go [hunk 14]
```diff
@@ -0,0 +1,98 @@
...
+
+	return zeekDNS.Logs()
+}
+
+// LogType returns the log type supported by this parser
+func (p *ZeekDNSParser) LogType() string {
+	return "Zeek.DNS"
+}
+
+func (event *ZeekDNS) updatePantherFields(p *ZeekDNSParser) {
```
Comment (kostaspap) [inline internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: According to the description the Query field seems to contain a domain? `The domain name that is the subject of the DNS query.`. We should add it to the AnyDomainNames

Comment (rleighton) [reply internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: Yes! I could have sworn I asked for this previously.

Comment (rleighton) [reply internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: Also, the answers field should be processed , checked if they are ips and put into IP and else put into domains.

Comment (sardev08) [reply internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: thanks for a comments, fixed

File: internal/log_analysis/log_processor/parsers/zeeklogs/dns.go [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+
+	"github.com/panther-labs/panther/internal/log_analysis/log_processor/parsers"
+	"github.com/panther-labs/panther/internal/log_analysis/log_processor/parsers/timestamp"
+)
+
+var ZeekDNSDesc = `Zeek DNS activity
+Reference: https://docs.zeek.org/en/current/scripts/base/protocols/dns/main.zeek.html#type-DNS::Info`
+
+// nolint:lll
+type ZeekDNS struct {
```
Comment (rleighton) [inline internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: You are missing  
```
qclass: count&log&optional
The QCLASS value specifying the class of the query.

qclass_name: string&log&optional
A descriptive name for the class of the query.

qtype: count&log&optional
A QTYPE value specifying the type of the query.

qtype_name: string&log&optional
A descriptive name for the type of the query.
```

File: internal/log_analysis/log_processor/parsers/zeeklogs/dns.go [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+
+func (event *ZeekDNS) updatePantherFields(p *ZeekDNSParser) {
+	event.SetCoreFields(p.LogType(), (*timestamp.RFC3339)(event.Ts), event)
+	if net.ParseIP(*event.IDOrigH) != nil {
+		event.AppendAnyIPAddresses(*event.IDOrigH)
+	}
+	if net.ParseIP(*event.IDRespH) != nil {
+		event.AppendAnyIPAddresses(*event.IDRespH)
+	}
+	if event.Query != nil {
```
Comment (rleighton) [inline internal/log_analysis/log_processor/parsers/zeeklogs/dns.go]: You have to check qtype and only do this on A and AAA records.

**Comment by nhakmiller** [comment]
@all-contributors add @sardev08 for code.
